src/gamescope/wayland/metrics/dbus.rs

```diff
@@ -1,39 +1,75 @@
-use std::time::SystemTime;
-
-use gamescope_wayland_client::mangoapp::{
-    message::MangoAppMsgV1,
-    queue::{MangoAppMsgQueue, QueueError},
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::JoinHandle,
+    time::{Duration, SystemTime},
 };
+
+use gamescope_wayland_client::mangoapp::message::MangoAppMsgV1;
 use thiserror::Error;
+use tokio::time::{error::Elapsed, timeout};
 use zbus::{dbus_interface, fdo};
 
+use crate::gamescope::wayland::metrics::mango_app_reader::MangoAppMsgQueueReader;
+
 #[derive(Error, Debug)]
 pub enum MetricsError {
     #[error("failed to read metrics message from gamescope: {0}")]
-    QueueError(#[from] QueueError),
+    QueueError(#[from] tokio::sync::watch::error::RecvError),
+    #[error("timedout trying to read gamescope queue message: {0}")]
+    QueueTimeout(#[from] Elapsed),
 }
 
 impl From<MetricsError> for fdo::Error {
     fn from(value: MetricsError) -> Self {
         match value {
             MetricsError::QueueError(queue_error) => fdo::Error::Failed(queue_error.to_string()),
+            MetricsError::QueueTimeout(queue_error) => fdo::Error::Failed(queue_error.to_string()),
         }
     }
 }
 
 pub struct DBusInterface {
-    queue: MangoAppMsgQueue,
+    handle: Option<JoinHandle<()>>,
+    cancel_flag: Arc<AtomicBool>,
     state: Option<MangoAppMsgV1>,
     last_update: SystemTime,
+    request_tx: std::sync::mpsc::Sender<()>,
+    response_rx: tokio::sync::watch::Receiver<Option<MangoAppMsgV1>>,
+}
+
+impl Drop for DBusInterface {
+    fn drop(&mut self) {
+        self.cancel_flag.store(true, Ordering::Relaxed);
+        let _ = self.request_tx.send(());
+
+        if let Some(handle) = self.handle.take() {
+            let _ = handle.join();
+        }
+    }
```

Comment:
Implementing `Drop` is not required if we just rely on looking at when the channel closes.

```suggestion
```